docs/proposals/volume-idempotency.md

```diff
@@ -0,0 +1,229 @@
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/proposals/volume-idempotency.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+## Volume plugins and idempotency
+
+Currently, volume plugins have a `SetUp` method which is called repeatedly in the kubelet sync loop.
+Implementers of plugins must implement idempotency checks in `SetUp` to avoid doing duplicate work.
+In order to simplify life for developers of plugins and correctly support idempotent application
+of ownership management to volumes, we should externalize these idempotency checks and perform them
+in the kubelet.
+
+### Types of idempotence checks
+
+There are two types of idempotence checks that volumes perform today:
+
+1.  Checking for the presence of a mount point
+2.  Checking for the presence of a marker file for the volume in the metadata directory for the
+    plugin type
+
+Externalized idempotency management must support both of these types of checks.  The marker file
+method requires that the marker file be created after `SetUp` is run successfully.
+
+### Non-idempotent volumes
+
+The hostPath volume does not implement an idempotency check because its `SetUp` method is a no-op.
+It should be possible to represent that a volume does not have any idempotency concerns.
+
+### Refreshable volume plugins
+
+Some volume plugins are designed to refresh, like the downward API volume plugin.  In order to
+decouple the implementation of refresh and SetUp logic, there should be a Refresh() method that
+plugins can implement and proper machinery to ensure that only plugins that want this need to
+implement it.
+
+### Volume plugins that wrap one another
+
+Some volume plugins wrap others:
+
+1.  Secrets, downward API, git repo are based on EmptyDir
+2.  PV claim plugin instantiates instances of the underlying volume plugin
+
+In order to correctly externalize idempotency management, the Kubelet must handle nested calls to
+`SetUp`; plugins must be able to invoke the externalized flow correctly for plugins they use
+transitively.
+
+## Ownership management and idempotency
+
+Ownership management of volumes must be idempotent for non-refreshable volumes.  It should run once
+and only once after `Setup` has run correctly.  If a volume that uses the 'check for a mountpoint'
+type of idempotency check, and the volume must be remounted, the ownership management should be
+reapplied.
+
+### Ownership management and refreshable volumes
+
+For volume plugins that are refreshable, ownership management must be reapplied if the refresh logic
+makes changes to the state of the volume.  For example, if data changes in the downward API volume,
+the ownership management must be re-applied to that volume.
```

Comment:
Hmm.. what is the use case here ? My intuition is that ownership management should be applied to all volumes only once.